hrjobcontract/CRM/Hrjobcontract/Upgrader.php

```diff
@@ -1065,41 +1065,91 @@ public function createInsurancePlanTypes() {
   }
 
   /**
-   * Changes Pension Provider to be a contact, instead of an option_value.
+   * Changes Pension Provider to be a contact, instead of an option_value.  All
+   * required changes are wrapped in a transaction and rolled-back if any 
+   * problems arise.
    * 
-   * @return true
+   * @return boolean
+   *   true on success
+   * 
+   * @throws Exception
    */
   function upgrade_1026() {
     $messages = [];
     $pensionsTableName = CRM_Hrjobcontract_BAO_HRJobPension::getTableName();
 
-    $this->createPensionProviderType($messages);
-    $this->replacePensionProviders($pensionsTableName, $messages);
-    $this->removePensionTypeOptionGroup($messages);
-
-    CRM_Core_DAO::executeQuery("
-      ALTER TABLE {$pensionsTableName} 
-      MODIFY pension_type INT(10) unsigned,
-      ADD CONSTRAINT pension_type_contact_id_fk
-      FOREIGN KEY(pension_type)
-      REFERENCES civicrm_contact(id)
-      ON DELETE SET NULL ON UPDATE CASCADE;
-    ");
+    $tx = new CRM_Core_Transaction();
+    try {
+      $this->createPensionProviderType($messages);
+      $this->replacePensionProviders($pensionsTableName, $messages);
+      $this->removePensionTypeOptionGroup($messages);
+      $this->alterPensionsTable($pensionsTableName, $messages);
+    } catch (Exception $ex) {
+      $tx->rollback();
+      throw new Exception('Cannot do upgrade - ' . $ex->getMessage());
+    }
 
     if (!empty($messages) && function_exists('drupal_set_message')) {
```

Comment:
Actually, I was using messages to notify about minor incidents, for example, trying to create a contact type that already exists.  The API will throw an exception, but I capture the exception and let it pass, so long as there is a contact type with 'Pension_Provider' name.  I know the probability of someone creating a "Pension_Provider" contact type before running this upgrade is slim, but then again, should the upgrade fail if this happens? Same idea applies for the rest of the messages: if an exception is thrown on an API call, but I verify everything is as it's supposed to be, I use the $messages array to warn, but let the upgrade finish succesfully.